aspnet/security/cross-site-scripting.rst

```diff
@@ -1,8 +1,256 @@
+    document.write("<br />")
+    document.write(clientSideUntrustedInputHtml5);
+  </script>
+
+This will produce the following HTML
+
+.. code-block:: html
+
+  <div
+    id="injectedData"
+    data-untrustedinput="&lt;&quot;123&quot;&gt;" />
+
+  <script>
+    var injectedData = document.getElementById("injectedData");
+
+    var clientSideUntrustedInputOldStyle =
+        injectedData.getAttribute("data-untrustedinput");
+
+    var clientSideUntrustedInputHtml5 =
+        injectedData.dataset.untrustedinput;
+
+    document.write(clientSideUntrustedInputOldStyle);
+    document.write("<br />")
+    document.write(clientSideUntrustedInputHtml5);
+  </script>
+
+Which, when it runs, will render the following;
+
+.. code-block
+
+  <"123">
+  <"123">
+
+You can also call the JavaScript encoder directly,
+
+.. code-block
+
+  @using System.Text.Encodings.Web;
+  @inject JavaScriptEncoder encoder;
+
+  @{
+      var untrustedInput = "<\"123\">";
+  }
+
+  <script>
+      document.write("@encoder.Encode(untrustedInput)");
+  </script>
+
+This will render in the browser as follows;
+
+.. code-block:: html
+
+  <script>
+      document.write("\u003C\u0022123\u0022\u003E");
+  </script>
+
+.. WARNING::  
+  Do not concatenate untrusted input in JavaScript to create DOM elements. You should use ``createElement()`` and assign 
+  property values appropriately such as ``node.TextContent=``, or use ``element.SetAttribute()`/`element[attribute]=`` 
+  otherwise you expose yourself to DOM-based XSS.
+
+Accessing encoders in code
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The HTML, JavaScript and URL encoders are available to your code in two ways, you can inject them via :ref:`dependency injection <fundamentals-dependency-injection>` 
+or you can use the default encoders contained in the ``System.Text.Encodings.Web`` namespace. If you use the default encoders then 
+any :ref:`customization <security-cross-site-scripting-customization>` you applied to character ranges to be treated as safe will not take effect - 
+the default encoders use the safest encoding rules possible.
+
+To use the configurable encoders via DI your constructors should take an `HtmlEncoder`, `JavaScriptEncoder` and `UrlEncoder` parameter as appropriate. For example;
+
+.. code-block:: c#
+
+  public class HomeController : Controller
+  {
+      HtmlEncoder _htmlEncoder;
+      JavaScriptEncoder _javaScriptEncoder;
+      UrlEncoder _urlEncoder;
+
+      public HomeController(HtmlEncoder htmlEncoder, 
+                            JavaScriptEncoder javascriptEncoder, 
+                            UrlEncoder urlEncoder)
+      {
+          _htmlEncoder = htmlEncoder;
+          _javaScriptEncoder = javascriptEncoder;
+          _urlEncoder = urlEncoder;
+      }
+  }
+
+Encoding URL Parameters
+^^^^^^^^^^^^^^^^^^^^^^^
+
+If you want to build a URL query string with untrusted input as a value you should use the ``UrlEncoder`` to encode the value. For example,
```

Comment:
To build a URL query string with untrusted input as a value ~~you should~~ use the ``UrlEncoder`` to encode the value. For example:

We don't like should directives